### ISSUE TITLE
memory problem with the Julia integration of GAP and Polymake

### ISSUE THREAD

**Issue by ThomasBreuer** [oscar-system/GAP.jl#280]
Currently Klaus Lux is using our setup for certain computations, which means that one starts GAP (the master branch, using Julia's garbage collector, plus some unpublished GAP packages that consist only of interpreted GAP code) and loads its JuliaInterface package, which loads Polymake.jl into Julia and then uses the GAP-Julia interface to call (frequently) some Polymake functions. This works fine in most of the cases, but occasionally we run into the following error.
```
Error, Exception occured at Polymake side:
C++ object was deleted in
  Julia.Base.getproperty( obj, _JL_RNAM_TO_JULIA_SYMBOL( rnam )
 ) at /users/oscar/gap-master/gap/pkg/GAPJulia/JuliaInterface/gap/adapter.gi called from
res := lp.MILP.MINIMAL_VALUE; at /users/klux/pkg/moc/gap/mocjulia.g:52 called from
MinimalIntegralSolution( z, A, b ) at /users/klux/pkg/moc/gap/mocjulia.g:134 called from
SolveViaPolymake( z, A, bb ) at /users/klux/pkg/moc/gap/polymake.gi:338 called from
MOC_MinimalMultiplicityMFS( scp, PinPS, proj, multfrees[depth], nonpims, M, k
 ) at /users/klux/pkg/moc/gap/polymake.gi:583 called from
SubtractAndStrip_tree( proj - maxes[multfrees[depth]][1] * PinPS[PS[multfree[depth]]], maxes,
 depth + 1, k ); at /users/klux/pkg/moc/gap/polymake.gi:586 called from
...  at *stdin*:27
```
We have not been able to reproduce this problem.
It occurs from time to time after many successful calls,
and if we repeat such a failed GAP session then we do not run into this error at the same place.

(Background:
The object 'lp' is a polytope object in which 'MILP' has been set via a Julia function, and GAP calls 'lp.MILP.MINIMAL_VALUE'.
The '.' access in GAP is executed by a call to Julia's 'getproperty'.
Here I cannot decide which of the two '.' calls fails.)

**Comment by ThomasBreuer** [comment]
Here are the Julia and GAP interface code which are involved.
First the Julia module:
```julia
module MOCutils

# temporary workaround in order to prevent a Julia crash
ENV[ "OMP_NUM_THREADS" ] = 1

import Polymake

"""
    MatrixFromNestedArray( lst )
> Return a 2-dim array created from the 1-dim array of 1-dim arrays `lst`.
> (Note that GAP's `GAPToJulia` creates nested arrays.)
"""
function MatrixFromNestedArray( lst ) return copy( hcat( lst... )' ) end

raw"""
    PolytopeWithMixedIntegerLinearProgram( inequalities, objective, intvars )
> `inequalities` must be a 2-dim. array of integers
> with rows of the form `[ a_0, a_1, ..., a_n ]`
> meaning the condition a_0 + a_1 x_1 + ... + a_n x_n \geq 0.
> `objective` must be a 1-dim. array of integers z_1, z_2, ..., z_n
> meaning that z_1 x_1 + z_2 x_2 + ... + z_n x_n shall be optimized.
"""
function PolytopeWithMixedIntegerLinearProgram( inequalities, objective, intvars )
    p = Polymake.@pm Polytope.Polytope( :INEQUALITIES => inequalities )
    p.MILP = Polymake.@pm Polytope.MixedIntegerLinearProgram(
             :LINEAR_OBJECTIVE => objective,
             :INTEGER_VARIABLES => convert( Array{Int,1}, intvars ) )

    return p
end

# This function is needed only because macros cannot be called via the
# GAP to Julia interface.
function PolytopeByInequalities( inequalities )
    return Polymake.@pm Polytope.Polytope( :INEQUALITIES => inequalities )
end

end
```
And now the GAP code:
```GAP
#############################################################################
##
#F  MinimalIntegralSolution( <z>, <A>, <b> )
##
##  Let <z> be a vector of length $n$,
##  <A> be an $m \times n$ matrix,
##  and <b> be a vector of length $m$.
##
##  The function computes the minimum of the scalar product $<z> * x$,
##  taken over the integral solutions of the system of linear inequalities
##  $<A> * x \leq <b>$,
##  with $x_i \geq 0$ for all $1 \leq i \leq n$.
##  If no integral solution exists then 'fail' is returned,
##  otherwise the minimum is returned.
##
MinimalIntegralSolution:= function( z, A, b )
    local inequalities, n, mat, obj, intvars, lp, res, type, ratres;

    # Rewrite the input.
    inequalities:= List( [ 1 .. Length( A ) ],
                         i -> Concatenation( [ b[i] ], -A[i] ) );

    # Apparently zero rows are not allowed.
    # (Zero coefficients with nonzero r.h.s. are allowed.)
    inequalities:= Filtered( inequalities, x -> not IsZero( x ) );

    # Add the inequalities that force $x_i \geq 0$.
    # (The property 'POSITIVE' cannot  be forced.)
    n:= Length( z );
    Append( inequalities, IdentityMat( n+1 ){ [ 2 .. n+1 ] } );

    # Create the inputs for the Julia function.
    mat:= Julia.MOCutils.MatrixFromNestedArray( GAPToJulia(
              inequalities ) );
    obj:= GAPToJulia( JuliaEvalString( "Array{Int64,1}" ),
              Concatenation( [ 0 ], z ) );
    intvars:= GAPToJulia( [ 0 .. Length( z ) ] );

    # Call the Julia function.
    lp:= Julia.MOCutils.PolytopeWithMixedIntegerLinearProgram( mat, obj, intvars );

    # Workaround:
    # For the test, make sure that the polytope is not empty.
    if not lp.FEASIBLE then
      return fail;
    fi;

    # Compute the optimal solution.
    res:= lp.MILP.MINIMAL_VALUE;

    # Translate the solution to GAP.
    if Julia.Base.isnothing( res ) then
      # no solution
      return fail;
    else
      # Convert the Polymake object to a Julia object.
      type:= JuliaEvalString( "Rational{Int}" );
      ratres:= type( res );

      # Convert the Julia object to a GAP object.
      return JuliaToGAP( IsRat, ratres );
    fi;
end;


#############################################################################
##
#F  IsSolvableSystemOfInequalities( <A>, <b> )
##
##  Let <A> be an $m \times n$ matrix of integers
##  and <b> be a vector of integers, of length $m$.
##  The function returns 'true' if the system of linear inequalities
##  $<A> * x \leq <b>$ has a solution $x$ that consists of nonnegative
##  rationals, and 'false' otherwise.
##
IsSolvableSystemOfInequalities:= function( A, b )
    local inequalities, n, mat;

    # Rewrite the input.
    inequalities:= List( [ 1 .. Length( A ) ],
                         i -> Concatenation( [ b[i] ], -A[i] ) );

    # Apparently zero rows are not allowed.
    # (Zero coefficients with nonzero r.h.s. are allowed.)
    inequalities:= Filtered( inequalities, x -> not IsZero( x ) );

    # Add the inequalities that force $x_i \geq 0$.
    # (The property 'POSITIVE' cannot  be forced.)
    n:= Length( A[1] );
    Append( inequalities, IdentityMat( n+1 ){ [ 2 .. n+1 ] } );

    # Create the inputs for the Julia function.
    mat:= Julia.MOCutils.MatrixFromNestedArray( GAPToJulia(
              inequalities ) );

    # Create the polytope, and check whether it is empty.
    return Julia.MOCutils.PolytopeByInequalities( mat ).FEASIBLE;
end;
```

**Comment by kalmarek** [comment]
@ThomasBreuer this looks very clean to the point I wonder if the problem might be on the polymake(.jl) side.

could you provide a sample ` mat`, `obj` and `intvars`? (easy, but not trivial example)

If i understand correctly `IsSolvableSystemOfInequalities` is superfluous here?

**Comment by sebasguts** [comment]
@kalmarek I do not believe in a Polymake.jl issue, but rather that the PR @fingolfin and I provided solved this.

Nevertheless, @ThomasBreuer, if you would have an example to reproduce this, it would be great to have a test for this.

**Comment by ThomasBreuer** [comment]
@kalmarek, yes, `IsSolvableSystemOfInequalities` does not show up in the error case.
I had included it for the sake of completeness, because it is the only other function besides `MinimalIntegralSolution` that uses Julia and occurs in the computations.

I do not have example code to reproduce the error case, currently we are testing the fix proposed by @sebasguts; this will take some time.

**Comment by kalmarek** [comment]
@ThomasBreuer I understand that reproducing error is hard and I don't ask for such an example.

Any (non-trivial) example input to call the code would do, so that I can generate more examples alike and repetitively run them putting some stress on the GC and julia ↔ polymake interface.

**Comment by mohamed-barakat** [comment]
Thank you very much @sebasguts and @fingolfin.

**Comment by ThomasBreuer** [comment]
@kalmarek o.k., here is one example; in one of the error cases, it was the one to be processed.
```
mat = [
1 -1 0 0 0 0 0 0 0; 1 0 -1 0 0 0 0 0 0; 1 0 0 -1 0 0 0 0 0;
1 0 0 0 -1 0 0 0 0; 1 0 0 0 0 -1 0 0 0; 1 0 0 0 0 0 -1 0 0; 
1 0 0 0 0 0 0 -1 0; 1 0 0 0 0 0 0 0 -1; 1 -1 0 0 0 0 0 0 0; 
0 1 0 0 0 0 0 0 0; 1 0 -1 0 0 0 0 0 0; 0 0 1 0 0 0 0 0 0; 
1 0 0 -1 0 0 0 0 0; 0 0 0 1 0 0 0 0 0; 1 0 0 0 -1 0 0 0 0; 
0 0 0 0 1 0 0 0 0; 1 0 0 0 0 -1 0 0 0; 0 0 0 0 0 1 0 0 0; 
1 0 0 0 0 0 -1 0 0; 0 0 0 0 0 0 1 0 0; 1 0 0 0 0 0 0 -1 0;
0 0 0 0 0 0 0 1 0; 1 0 0 0 0 0 0 0 -1; 0 0 0 0 0 0 0 0 1; 
-1 1 0 0 0 0 0 0 0; 1 -1 0 0 0 0 0 0 0; 2 0 -1 -1 0 0 0 0 0; 
-1 0 1 1 0 0 0 0 0; -1 1 1 -1 0 0 0 0 0; 1 -1 -1 1 0 0 0 0 0; 
0 0 1 0 -1 0 0 0 0; 0 0 -1 0 1 0 0 0 0; 1 -1 0 1 0 -1 0 0 0; 
-1 1 0 -1 0 1 0 0 0; 1 0 0 0 -1 0 0 0 0; 0 0 0 0 1 0 0 0 0;
1 0 0 0 0 -1 0 0 0; 0 0 0 0 0 1 0 0 0; 0 0 -1 0 1 0 0 0 0; 
0 0 1 0 -1 0 0 0 0; 1 0 -2 0 1 0 0 0 0; 0 0 2 0 -1 0 0 0 0;
2 -1 -1 -1 1 0 0 0 0; -1 1 1 1 -1 0 0 0 0; 0 -1 0 0 0 1 0 0 0; 
0 1 0 0 0 -1 0 0 0; 2 -2 -1 1 1 -1 0 0 0; -1 2 1 -1 -1 1 0 0 0; 
2 -2 -1 1 0 0 0 0 0; -1 2 1 -1 0 0 0 0 0; 4 -4 -1 1 0 0 0 0 0;
-2 4 1 -1 0 0 0 0 0; 5 -6 -3 2 1 1 0 0 0; -3 6 3 -2 -1 -1 0 0 0; 
4 -2 -1 1 0 -2 0 0 0; -1 2 1 -1 0 2 0 0 0; 1 0 -1 0 1 0 0 -1 0;
0 0 1 0 -1 0 0 1 0; 5 2 2 0 -3 -4 -1 -1 0; 1 -2 -2 0 3 4 1 1 0;
3 -2 -3 0 3 -1 0 0 0; -2 2 3 0 -3 1 0 0 0; 2 -4 -2 2 1 1 0 0 0; 
0 4 2 -2 -1 -1 0 0 0; 0 2 1 -1 -1 0 0 -1 0; 1 -2 -1 1 1 0 0 1 0; 
0 2 1 -1 -1 0 0 -1 0; 1 -2 -1 1 1 0 0 1 0; 1 1 0 -1 0 0 0 -1 0; 
0 -1 0 1 0 0 0 1 0; 1 0 0 0 0 0 0 -1 0; 0 0 0 0 0 0 0 1 0;
1 0 0 0 0 0 0 -1 0; 0 0 0 0 0 0 0 1 0; 1 1 0 0 0 -1 0 -1 0;
0 -1 0 0 0 1 0 1 0; 4 2 3 -1 -4 -2 -1 -1 0; 1 -2 -3 1 4 2 1 1 0;
4 -1 -1 1 0 -1 0 -2 0; -2 1 1 -1 0 1 0 2 0; 1 1 0 -2 0 1 1 -1 -1;
1 -1 0 2 0 -1 -1 1 1; 21 0 -6 -1 4 -3 0 -13 -2; -2 0 6 1 -4 3 0 13 2;
21 -2 -5 1 4 -4 0 -13 -2; -2 2 5 -1 -4 4 0 13 2; 14 3 -1 -3 0 -2 1 -9 -3;
0 -3 1 3 0 2 -1 9 3; 18 -5 -4 2 3 -2 0 -9 -3; -3 5 4 -2 -3 2 0 9 3;
18 3 -3 -4 1 -1 1 -12 -3; 1 -3 3 4 -1 1 -1 12 3; 15 10 2 -9 -3 1 2 -14 -4;
5 -10 -2 9 3 -1 -2 14 4; 28 -6 -6 -2 3 1 2 -14 -6; -3 6 6 2 -3 -1 -2 14 6;
26 16 0 -13 -1 -1 3 -24 -6; 6 -16 0 13 1 1 -3 24 6; 37 -5 -6 -1 5 -3 1 -20 -8;
-2 5 6 1 -5 3 -1 20 8; 28 12 -1 -15 -1 4 5 -23 -9; 6 -12 1 15 1 -4 -5 23 9;
44 24 0 -23 -1 2 6 -41 -11; 10 -24 0 23 1 -2 -6 41 11; 0 1 0 0 0 0 0 0 0; 
0 0 1 0 0 0 0 0 0; 0 0 0 1 0 0 0 0 0; 0 0 0 0 1 0 0 0 0; 
0 0 0 0 0 1 0 0 0; 0 0 0 0 0 0 1 0 0; 0 0 0 0 0 0 0 1 0; 
0 0 0 0 0 0 0 0 1];

obj = [0, 0, 6, 0, -4, 3, 0, 13, 2]

intvars = [0, 1, 2, 3, 4, 5, 6, 7, 8]
```

**Comment by kalmarek** [comment]
I left it fuzzing julia ↔ Polymake for a night with small randomness in `mat`, and while I didn't experience any crush I did find an example which is feasible, but `MINIMAL_VALUE` is still nothing:
```
mat = [1 -1 0 0 0 0 0 0 0; 1 0 -1 0 0 0 0 0 0; 1 0 0 -1 0 0 0 0 0; 1 0 0 0 -1 0 0 0 0; 1 0 0 0 0 -1 0 0 0; 1 0 0 1 0 0 -1 0 0; 1 0 0 0 0 0 0 -1 0; 1 0 0 0 0 0 0 0 -1; 1 -1 0 0 0 0 0 0 0; 0 1 0 0 0 0 0 0 0; 1 0 -1 0 0 0 0 0 0; 0 0 1 0 0 0 0 0 0; 1 0 0 -1 0 0 0 0 0; 0 0 0 1 0 0 0 0 0; 1 0 0 0 -1 0 -1 0 0; 0 0 0 0 1 0 0 0 0; 1 0 0 0 0 -1 0 0 0; 0 0 0 0 0 1 0 0 0; 1 0 0 0 0 0 -1 0 0; 0 0 0 0 0 0 1 0 0; 1 0 0 0 0 0 0 -1 0; 0 0 0 0 0 0 0 1 0; 1 0 0 0 0 0 0 0 -3; 0 0 0 0 0 0 0 0 1; -1 1 0 0 2 0 0 0 0; 1 -1 0 0 0 0 0 0 0; 2 0 -1 -1 0 0 0 0 0; -1 0 1 1 0 0 0 0 0; -1 1 1 -1 0 0 0 0 0; 1 -1 -1 1 0 0 0 0 0; 0 0 1 0 -1 0 0 0 0; 0 0 -1 0 1 0 0 0 0; 1 -1 0 1 0 -1 0 0 0; -1 1 0 -1 0 1 0 0 0; 1 0 0 0 -1 0 0 0 0; 0 0 0 0 1 0 0 0 0; 1 0 0 0 0 -1 0 0 0; 0 0 0 0 0 1 0 0 0; 0 0 -1 0 1 0 0 0 0; 0 0 1 0 -1 0 0 0 0; 1 0 -2 0 1 0 0 0 0; 0 0 2 0 -1 0 0 0 0; 2 -1 -1 -1 1 0 0 0 0; -1 1 1 1 -1 0 0 0 0; 0 -1 0 0 0 1 0 0 0; 0 1 0 0 0 -1 0 0 0; 2 -2 -1 1 1 -1 0 0 0; -1 2 1 -1 -1 1 0 0 0; 2 -2 -1 1 0 0 0 0 0; -1 2 1 -1 0 0 0 0 0; 4 -4 -1 0 0 0 0 0 0; -2 4 1 -1 0 0 0 0 0; 5 -6 -3 2 1 1 0 0 0; -3 6 3 -2 -1 -1 0 0 0; 4 -2 -1 1 0 -2 0 0 -1; -1 2 1 -1 0 2 0 0 0; 1 0 -1 0 1 0 0 -1 0; 0 0 1 0 -1 0 0 1 0; 5 2 2 0 -3 -4 -1 -1 0; 1 -2 -2 0 3 4 1 1 0; 3 -2 -3 0 3 -1 0 0 0; -2 2 3 0 -3 1 0 0 0; 2 -4 -2 2 1 1 0 0 0; 0 4 2 -2 -1 -1 0 0 0; 0 2 1 -1 -1 0 0 -1 0; 1 -2 -1 1 1 0 0 1 0; 0 2 1 -1 -1 0 0 -1 0; 1 -2 -1 1 1 0 0 1 0; 1 1 0 -1 0 0 0 -1 0; 0 -1 0 1 0 0 0 1 0; 1 0 0 0 0 0 0 -1 0; 0 0 0 0 0 0 0 1 0; 1 0 0 0 0 0 0 -1 0; 0 0 0 0 0 0 0 1 0; 1 1 0 0 0 -1 0 -1 0; 0 -1 0 0 0 1 0 1 0; 4 2 3 -1 -4 -2 -1 -1 0; 1 -2 -3 1 4 2 1 1 0; 4 -1 -1 1 0 -1 0 -2 0; -2 1 1 -1 0 1 0 2 0; 1 1 0 -2 0 1 1 -1 -1; 1 -1 0 2 0 -1 -1 1 1; 21 0 -6 -1 4 -3 0 -13 -2; -2 0 6 1 -4 3 0 13 2; 21 -2 -5 1 4 -4 0 -13 -2; -2 2 5 -1 -4 4 0 13 2; 14 3 -1 -3 0 -2 1 -9 -3; 0 -3 1 3 0 2 -1 9 3; 18 -5 -4 2 3 -2 0 -9 -3; -3 5 4 -3 -3 2 0 9 3; 18 3 -3 -4 1 -1 1 -12 -3; 1 -3 3 4 -1 1 -1 12 3; 15 10 2 -9 -3 1 2 -14 -4; 5 -10 -2 9 3 -1 -2 14 4; 28 -6 -6 -2 3 1 2 -14 -6; -3 6 6 2 -3 -1 -2 14 6; 26 16 0 -13 -1 -1 3 -24 -6; 6 -16 0 13 1 1 -3 24 6; 37 -5 -6 -1 5 -3 1 -20 -8; -2 5 6 1 -5 3 -1 20 8; 28 12 -1 -15 -1 4 5 -23 -9; 6 -12 1 15 1 -4 -5 23 9; 44 24 0 -23 -1 2 6 -41 -11; 10 -24 0 23 1 -2 -6 41 11; 0 1 0 0 0 0 0 0 0; 0 0 1 0 0 0 0 0 0; 0 0 0 1 0 0 0 0 0; 0 0 0 0 1 0 0 0 0; 0 0 0 0 0 1 0 0 0; 0 0 0 0 0 0 1 0 0; 0 0 0 0 0 0 0 1 0; 0 0 0 0 0 0 0 0 1];
obj = [0, 0, 6, 0, -4, 3, 0, 13, 2];
intvars = [0, 1, 2, 3, 4, 5, 6, 7, 8];

P = PolytopeWithMixedIntegerLinearProgram(mat, obj, intvars);
@show P.FEASIBLE # true
@show isnothing(P.MILP.MINIMAL_VALUE) # true
```

I know that investigating such a large example is rather hard, but maybe something more can be said @lkastner?
(polymake was build from yesterdays master)

**Comment by fingolfin** [comment]
Oops, sorry, did not realize that merging PR #281 would close this issue. Of course we are still waiting for confirmation that the issue is fixed (and if it is not, we'll search for further potential causes).

**Comment by ThomasBreuer** [comment]
@kalmarek I do not catch the point.
An example of a feasible (nonempty) polytope without lattice points (one of my tests) is as follows.
```Julia
ENV[ "OMP_NUM_THREADS" ] = 1;
import Polymake

inequalities = [3 2 -3; 0 -2 3; 9 -2 -3; -6 2 3; 0 1 0; 0 0 1];
objective = [0, 0, 0];
intvars = [0, 1, 2];

p = Polymake.@pm Polytope.Polytope( :INEQUALITIES => inequalities );
p.MILP = Polymake.@pm Polytope.MixedIntegerLinearProgram(
             :LINEAR_OBJECTIVE => objective,
             :INTEGER_VARIABLES => intvars );

p.FEASIBLE  # true

typeof( p.MILP.MINIMAL_VALUE )  # Nothing
```

**Comment by kalmarek** [comment]
@ThomasBreuer I see, I thought that "feasibility" of MILP means that the polytope contains integer lattice points.

**Comment by lkastner** [comment]
@kalmarek I have not looked at your example, but maybe for clarification: The 'M' in `MILP` stands for mixed. This means that some variables are required to be integer, some can be rationals (or whatever). You can determine this by specifying `INTEGER_VARIABLES`. If this is left empty, then you are back at a normal `LP`. (`MILP` however then defaults to requiring all variables to be integers, since you should use a different solver for an `LP`) The polytope giving the feasible region is independent of the `INTEGER_VARIABLES`. Requiring some variables to be integers is not something that will result in a polytope. Just like the `LINEAR_OBJECTIVE`, this property is irrelevant for the feasible region.

Have a look at https://polymake.org/doku.php/documentation/master/polytope/mixedintegerlinearprogram . `MILP` is attached to a polytope (the feasible region). You only get the implication that if the underlying polytope is infeasible, there is (should be) no solution to the MILP. The other direction is false.

**Comment by kalmarek** [comment]
@lkastner there is no need for you to look at it. indeed `FEASIBLE` is the attribute of  polytope here, and not of `MILP`, so that settles it.

**Comment by fingolfin** [comment]
@ThomasBreuer did the problem occur again so far?

**Comment by ThomasBreuer** [comment]
@fingolfin With the fix from pull request #281, we did not run into this memory problem. I think this issue can be closed.
